src/mlpack/methods/ne/genome.hpp

```diff
@@ -0,0 +1,334 @@
+  void SortLinkGenes() {
+    struct DepthAndLink
+    {
+      double depth;
+      LinkGene link;
+ 
+      DepthAndLink(double d, LinkGene& l) : depth(d), link(l) {}
+
+      bool operator < (const DepthAndLink& dL) const
+      {
+        return (depth < dL.depth);
+      }
+    };
+
+    std::vector<double> toNeuronDepths;
+    for (int i=0; i<aLinkGenes.size(); ++i) {
+      NeuronGene toNeuron;
+      GetNeuronById(aLinkGenes[i].ToNeuronId(), toNeuron);
+      toNeuronDepths.push_back(toNeuron.Depth());
+    }
+
+    std::vector<DepthAndLink> depthAndLinks;
+    int linkGenesSize = aLinkGenes.size();
+    for (int i=0; i<linkGenesSize; ++i) {
+      depthAndLinks.push_back(DepthAndLink(toNeuronDepths[i], aLinkGenes[i]));
+    }
+
+    std::sort(depthAndLinks.begin(), depthAndLinks.end());
+
+    for (int i=0; i<linkGenesSize; ++i) {
+      aLinkGenes[i] = depthAndLinks[i].link;
+    }
+  }
+
+  // Activate genome. The last dimension of input is always 1 for bias. 0 means no bias.
+  void Activate(std::vector<double>& input) {
+    assert(input.size() == aNumInput);
+
+    SortLinkGenes();
+    
+    // Set all neurons' input to be 0.
+    for (int i=0; i<NumNeuron(); ++i) {
+      aNeuronGenes[i].Input(0);
+    }
+    
+    // Set input neurons.
+    for (int i=0; i<aNumInput; ++i) {
+      aNeuronGenes[i].Input(input[i]);  // assume INPUT, BIAS, OUTPUT, HIDDEN sequence
+      aNeuronGenes[i].Activation(input[i]);
+    }
+
+    // Activate hidden and output neurons.
+    for (int i = 0; i < NumLink(); ++i) {
+      if (aLinkGenes[i].Enabled()) {
+        int toNeuronIdx = GetNeuronIndex(aLinkGenes[i].ToNeuronId());
+        int fromNeuronIdx = GetNeuronIndex(aLinkGenes[i].FromNeuronId());
+        double input = aNeuronGenes[toNeuronIdx].Input() + 
+                       aNeuronGenes[fromNeuronIdx].Activation() * aLinkGenes[i].Weight();
+        aNeuronGenes[toNeuronIdx].Input(input);
+        
+        if (i == NumLink() - 1) {
+          aNeuronGenes[toNeuronIdx].CalcActivation();
+        } else if (GetNeuronIndex(aLinkGenes[i + 1].ToNeuronId()) != toNeuronIdx) {
+          aNeuronGenes[toNeuronIdx].CalcActivation();
+        }
+      }
+    }
+  }
+
+  // Get output vector.
+  void Output(std::vector<double>& output) {
+    output.clear();
+    for (int i=0; i<aNumOutput; ++i) {
+      output.push_back(aNeuronGenes[aNumInput + i].Activation());
+    }
+  }
+
+  // Set random link weights between [lo, hi].
+  void RandomizeWeights(const double lo, const double hi) {
+    for (int i=0; i<aLinkGenes.size(); ++i) {
+      double weight = mlpack::math::Random(lo, hi);
+      aLinkGenes[i].Weight(weight); 
+    }
+  }
+
+  // Add link.
+  void AddLink(LinkGene& linkGene) {
+    aLinkGenes.push_back(linkGene);
+  }
+
+  // Add neuron.
+  void AddHiddenNeuron(NeuronGene& neuronGene) {
+    if (neuronGene.Type() == HIDDEN) {
+      aNeuronGenes.push_back(neuronGene);
+    }
+  }
+
+  // Show genome structure.
+  void Show() {
```

Comment:
Every class you like to serialize should implement the Serialize, in case of the Genome I guess the important parameter you like to serialize is the list of genes and links and some other parameter like fitness:
```
template<typename Archive>
void Serialize(Archive& ar, const unsigned int /* version */)
{
  ar & data::CreateNVP(neuronGenes, "neuronGenes");
  ar & data::CreateNVP(linkGenes, "linkGenes");
  ar & data::CreateNVP(fitness, "fitness");
}
```

You can save a model (Genome) using:

```
Genome genome;
data::Save("output_model_file.xml", "genome_model", genome);
```

Loading is also simple:

```
Genome genome;
data::Load("input_model_file.xml"), "genome_model", genome);
```

Also you might find: https://github.com/mlpack/mlpack/wiki/DesignGuidelines#requirements-of-each-class helpful